File: Assets/src/model/grasses/Agave.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Walk over it to harvest.")]
public class Agave : Grass, IActorEnterHandler {
  public static bool CanOccupy(Tile tile) => Mushroom.CanOccupy(tile);
  public Agave(Vector2Int pos) : base(pos) {}

  public void HandleActorEnter(Actor actor) {
    var player = GameModel.main.player;
    if (actor == player) {
      this.BecomeItemInInventory(new ItemAgave(1));
    }
  }
}

[Serializable]
[ObjectInfo("agave", flavorText: "An unassuming and earthy succulent with thick, nutrient bearing leaves.")]
class ItemAgave : Item, IStackable {
  public ItemAgave(int stacks) {
    this.stacks = stacks;
  }

  public int stacksMax => 4;
  private int _stacks;
  public int stacks {
    get => _stacks;
    set {
      if (value < 0) {
        throw new ArgumentException("Setting negative stack!" + this + " to " + value);
      }
      _stacks = value;
      if (_stacks == 0) {
        Destroy();
      }
    }
  }

  public void Refine(Player player) {
    player.SetTasks(new GenericPlayerTask(player, () => RefineImpl(player)));
  }

  private void RefineImpl(Player player) {
    if (stacks < stacksMax) {
      throw new CannotPerformActionException("Gather 4 stacks of Agave first.");
    }
    player.floor.Put(new ItemOnGround(player.pos, new ItemAgaveHoney(), player.pos));
    Destroy();
  }

  public override List<MethodInfo> GetAvailableMethods(Player player) {
    var methods = base.GetAvailableMethods(player);
    methods.Add(GetType().GetMethod("Refine"));
    return methods;
  }

  internal override string GetStats() => $"Gather 4 to Refine into Honey.\nHoney heals 1 HP and removes all Debuffs (red outlined Statuses).";
}

[Serializable]
[ObjectInfo("agave-honey", flavorText: "A restorative and tasty treat!")]
class ItemAgaveHoney : Item, IEdible {
  public void Eat(Actor a) {
    a.Heal(1);
    var debuffs = a.statuses.list.Where((s) => s.isDebuff);
    if (debuffs.Count() > 0) {
      foreach (var debuff in debuffs) {
        debuff.Remove();
      }
      GameModel.main.DrainEventQueue();
    }
    Destroy();
  }

  internal override string GetStats() => "Heals 1 HP and removes all Debuffs (red outlined Statuses).";
}

File: web/src/model/grasses/Agave.ts
import { Grass } from './Grass';
import { ACTOR_ENTER_HANDLER, type IActorEnterHandler } from '../../core/types';
import { GameModelRef } from '../GameModelRef';
import { Item, STACKABLE_TAG, type IStackable } from '../Item';
import { EDIBLE_TAG, type IEdible } from '../Item';
import { ItemOnGround } from '../ItemOnGround';
import { Mushroom } from './Mushroom';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Vector2Int } from '../../core/Vector2Int';
import type { Tile } from '../Tile';
import type { Actor } from '../Actor';

/**
 * Walk over to harvest. Collect 4 to refine into Agave Honey.
 * Port of C# Agave from Agave.cs.
 */
export class Agave extends Grass implements IActorEnterHandler {
  readonly [ACTOR_ENTER_HANDLER] = true as const;

  static canOccupy(tile: Tile): boolean {
    return Mushroom.canOccupy(tile);
  }

  constructor(pos: Vector2Int) {
    super(pos);
  }

  handleActorEnter(actor: any): void {
    const player = GameModelRef.main.player;
    if (actor === player) {
      this.becomeItemInInventory(new ItemAgave(1));
    }
  }

  private becomeItemInInventory(item: Item): void {
    const player = GameModelRef.main.player;
    const floor = this.floor;
    this.kill(player);
    if (!player.inventory.addItem(item, this)) {
      if (floor) {
        floor.put(new ItemOnGround(this.pos, item, this.pos));
      }
    }
  }
}

/**
 * Stackable item. Refine at 4 stacks to get Agave Honey.
 * Port of C# ItemAgave from Agave.cs.
 */
export class ItemAgave extends Item implements IStackable {
  readonly [STACKABLE_TAG] = true as const;

  private _stacks: number;
  get stacks(): number { return this._stacks; }
  set stacks(value: number) {
    if (value < 0) throw new Error('Negative stacks');
    this._stacks = value;
    if (this._stacks === 0) this.Destroy();
  }
  get stacksMax(): number { return 4; }

  constructor(stacks: number) {
    super();
    this._stacks = stacks;
  }

  getAvailableMethods(): string[] {
    const methods = super.getAvailableMethods();
    if (this.stacks >= this.stacksMax) {
      methods.push('Refine');
    }
    return methods;
  }

  refine(): void {
    if (this.stacks < this.stacksMax) return;
    const player = GameModelRef.main.player;
    player.floor.put(new ItemOnGround(player.pos, new ItemAgaveHoney(), player.pos));
    this.Destroy();
  }
}

/**
 * Edible item. Heals 1 HP and removes all debuffs.
 * Port of C# ItemAgaveHoney from Agave.cs.
 */
export class ItemAgaveHoney extends Item implements IEdible {
  readonly [EDIBLE_TAG] = true as const;

  eat(actor: Actor): void {
    actor.heal(1);
    const debuffs = [...actor.statuses.list].filter(s => s.isDebuff);
    for (const debuff of debuffs) {
      debuff.Remove();
    }
    if (debuffs.length > 0) {
      GameModelRef.main.drainEventQueue();
    }
    this.Destroy();
  }
}

entityRegistry.register('Agave', Agave);
